### DISCUSSION TITLE
Stream request cancellation

### DISCUSSION THREAD

**Discussion by Vrangz** [rwf2/Rocket#2552] [Questions] [answered]
Hi, is there a possibility to check if client aborted the stream?
I have built an endpoint which request some resource and stream it back to the requester, but the client can abort the request and I would like to detect it, so I can finish work earlier and offload the server.
The code may look like below:
```
#[post("/new-stream")]
pub fn my_func() -> EventStream<impl Stream<Item = Event>> {
    EventStream! {
        ...
        let mut stream = client.create_stream(request).await.unwrap();
        while let Some(result) = stream.next().await {
            if is_aborted() { 
                // close the stream and return
            }
            match result {
                Ok(response) => {
                    ...
                },
                Err(err) => {
                    ...
                }
            }
        }
    ...
    }
}
```

**Comment by SergioBenitez** [comment, marked as answer]
The `EventStream` _is_ the stream being written to the client. There is no bidrectional activity going on, so your example as written isn't particularly applicable or attainable. However, when the client disconnects, the stream is dropped, which is effectively the cancellation. The only way to detect this, using Rocket's provided stream implementations, is via a drop guard. I wrote an example of doing this here: https://github.com/SergioBenitez/Rocket/issues/2455#issuecomment-1483847660.

You could implement a stream type of your own that detects that the client has left and invokes a function, but that is outside the scope of the provided Rocket types. That would mean [handling this error condition](https://api.rocket.rs/v0.5-rc/src/rocket/response/stream/reader.rs.html#175). Alternatively, you could use a bidrectional stream, like [websockets](https://api.rocket.rs/master/rocket_ws/index.html), which does expose expected and unexpected connection closes.
